### PR TITLE
Add data points for each data read and generate duration distribution

### PR DESCRIPTION
**Pull request by twalluxio**
Add records of each file read, add histogram to get duration distribution for a file read job.

example: run the following command: `/opt/alluxio/bin/alluxio runClass alluxio.stress.cli.worker.StressWorkerBench --bench-timeout 10m --clients 8 --threads 8 --warmup 30s --duration 30s --buffer-size 1m --base alluxio:///stress-worker-base --cluster-limit 2 --cluster`, get result like 
[Example_results.txt](https://github.com/jiacheliu3/alluxio/files/12214376/Example_results.txt)

Part of the example result:

```
{
  ...
  "durationPercentiles" : [ 754, 902, 948, 986, 993, ..., 1779, 1842, 1889, 1999, 2087 ],
  ...
  "nodeResults" : {
    "Ec2Cluster-workers-1-1690800472497" : {
      ...
      "durationPercentiles" : [ 931, 992, 1006, 1035, 1051, ..., 1737, 1753, 1755, 1878, 1882 ],
      "dataPoints" : [ {
        "workerID" : "1",
        "threadID" : 32,
        "start" : 1690800581873,
        "duration" : 1373,
        "ioBytes" : 134217728,
        "startMs" : 1690800581873,
        "iobytes" : 134217728
      }, {
        "workerID" : "1",
        "threadID" : 32,
        "start" : 1690800583246,
        "duration" : 1274,
        "ioBytes" : 134217728,
        "startMs" : 1690800583246,
        "iobytes" : 134217728
      }, {
        ...
      }, {
        "workerID" : "1",
        "threadID" : 31,
        "start" : 1690800612060,
        "duration" : 1202,
        "ioBytes" : 134217728,
        "startMs" : 1690800612060,
        "iobytes" : 134217728
      } ],
      "className" : "alluxio.stress.worker.WorkerBenchTaskResult"
    },
    "Ec2Cluster-workers-0-1690800472496" : {
      ...
      "durationPercentiles" : [ 754, 795, 902, 938, 969, ..., 1889, 1939, 1999, 2032, 2087 ],
      "dataPoints" : [ {
        "workerID" : "0",
        "threadID" : 35,
        "start" : 1690800581558,
        "duration" : 1312,
        "ioBytes" : 134217728,
        "startMs" : 1690800581558,
        "iobytes" : 134217728
      }, {
        "workerID" : "0",
        "threadID" : 35,
        "start" : 1690800582870,
        "duration" : 1382,
        "ioBytes" : 134217728,
        "startMs" : 1690800582870,
        "iobytes" : 134217728
      }, {
        ...
      }, {
        "workerID" : "0",
        "threadID" : 34,
        "start" : 1690800611323,
        "duration" : 1069,
        "ioBytes" : 134217728,
        "startMs" : 1690800611323,
        "iobytes" : 134217728
      } ],
      "className" : "alluxio.stress.worker.WorkerBenchTaskResult"
    }
  },
  "className" : "alluxio.stress.worker.WorkerBenchSummary"
}
```

### CODE REVIEW

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchStats.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    public void setStart(long start) {
+        mStart = start;
+    }
+
+    public void setIOBytes(long ioBytes) {
+        mIOBytes = ioBytes;
+    }
+}
```
Comment (jiacheliu3) [inline dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchStats.java]: newline at the end of file

File: dora/stress/shell/src/main/java/alluxio/stress/cli/worker/StressWorkerBench.java [hunk 2]
```diff
@@ -123,16 +124,16 @@ public static void main(String[] args) {
...
-        "The test will create one file and repeatedly read the created file to test the "
-            + "performance",
-        "",
-        "Example:",
-        "# This would create a 100MB file with block size of 16KB and then read the file "
-            + "for 30s after 10s warmup",
-        "$ bin/alluxio runClass alluxio.stress.cli.worker.StressWorkerBench --clients 1 "
-            + "--base alluxio:///stress-worker-base --block-size 16k --file-size 100m "
-            + "--warmup 10s --duration 30s --cluster\n"
+            "A benchmarking tool to measure the read performance of alluxio workers in the cluster",
```
Comment (jiacheliu3) [inline dora/stress/shell/src/main/java/alluxio/stress/cli/worker/StressWorkerBench.java]: restore these extra whitespace

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java [hunk 3]
```diff
@@ -62,6 +65,7 @@ public void aggregateByWorker(WorkerBenchTaskResult result) {
     mParameters = result.mParameters;
 
     mRecordStartMs = result.mRecordStartMs;
+    // mRecordStartMs = Math.min(mRecordStartMs, result.mRecordStartMs);
```
Comment (jiacheliu3) [inline dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java]: why is this removed?

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java [hunk 4]
```diff
@@ -156,10 +160,22 @@ public void setErrors(List<String> errors) {
...
+
+  public List<WorkerBenchStats> getStats() {
+    return mStats;
+  }
+
+  public void appendStats(WorkerBenchStats stats) {
+    mStats.add(stats);
+  }
+
+  public void appendStats(List<WorkerBenchStats> stats) {
```
Comment (jiacheliu3) [inline dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java]:
```suggestion
  public void addDataPoints(Collection<WorkerBenchStats> stats) {
```

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchStats.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    @JsonProperty("workerID")
+    public String mWorkerID;
+    @JsonProperty("threadID")
+    public long mThreadID;
+
+    @JsonProperty("duration")
+    public long mDuration;
+    @JsonProperty("start")
+    public long mStart;
```
Comment (jiacheliu3) [inline dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchStats.java]:
```suggestion
    public long mStartMs;
```

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchStatsDeserializer.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+public class WorkerBenchStatsDeserializer extends JsonDeserializer<WorkerBenchStats> {
+
+    @Override
+    public WorkerBenchStats deserialize(JsonParser parser, DeserializationContext ctx)
+            throws IOException {
+        JsonNode node = parser.getCodec().readTree(parser);
+        return new WorkerBenchStats(
+                node.get("workerID").asText(), node.get("threadID").asLong(),
+                node.get("start").asLong(), node.get("duration").asLong(),
```
Comment (jiacheliu3) [inline dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchStatsDeserializer.java]:
```suggestion
                node.get("startMs").asLong(), node.get("duration").asLong(),
```

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchStats.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
+package alluxio.stress.worker;
```
Comment (jiacheliu3) [inline dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchStats.java]: missing license at the start of files

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchStats.java (hunk 5 above)
Comment (jiacheliu3) [reply dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchStats.java]: dont forget to update all getters/setters and json fields on the name change

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchSummary.java [hunk 8]
```diff
@@ -36,13 +39,15 @@ public final class WorkerBenchSummary extends GeneralBenchSummary<WorkerBenchTas
   private long mDurationMs;
   private long mEndTimeMs;
   private long mIOBytes;
+  private List<Long> mDurationPercentiles;
```
Comment (jiacheliu3) [inline dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchSummary.java:42]: is this ALL IO from ALL workers? Or just IO from ONE worker?

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java [hunk 9]
```diff
@@ -30,13 +30,15 @@ public final class WorkerBenchTaskResult implements TaskResult {
   private long mEndMs;
   private long mIOBytes;
   private List<String> mErrors;
+  private List<WorkerBenchStats> mStats;
```
Comment (jiacheliu3) [inline dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java]: to be more specific, let's call it `mDataPoints`

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchStats.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+
+@JsonDeserialize(using = WorkerBenchStatsDeserializer.class)
+public class WorkerBenchStats {
```
Comment (jiacheliu3) [inline dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchStats.java]: let's call this `WorkerOperationDataPoint`?

```
/**
 * One data point captures the information we collect from one I/O operation to a worker.
 * The one operation may be a full scan or positioned read on a file.
 */
@JsonDeserialize(using = WorkerOperationDataPointDeserializer.class)
class WorkerOperationDataPoint {...}
```

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java [hunk 11]
```diff
@@ -156,10 +160,22 @@ public void setErrors(List<String> errors) {
...
-    mErrors.add(errMesssage);
+  public void addErrorMessage(String errMessage) {
+    mErrors.add(errMessage);
+  }
+
+  public List<WorkerBenchStats> getStats() {
+    return mStats;
+  }
+
+  public void appendStats(WorkerBenchStats stats) {
```
Comment (jiacheliu3) [inline dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java]:
```suggestion
  public void addDataPoint(WorkerBenchStats stats) {
```

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Added a few more comments on naming, PTAL

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java (hunk 3 above)
Comment (twalluxio) [reply dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java]: originally the benchmark use `mRecordStartMs = result.mRecordStartMs;`, thinking about if necessary to change into `mRecordStartMs = Math.min(mRecordStartMs, result.mRecordStartMs);`

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchSummary.java (hunk 8 above)
Comment (twalluxio) [reply dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchSummary.java:42]: All IO from ALL workers I think. I haven't edited this variable.

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java (hunk 3 above)
Comment (jiacheliu3) [reply dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java]: oh good question, i don't really know exactly but I feel this is fine for now. You may print the `mRecordStartMs` for each result. If this change is necessary then you can add it in a later change.

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchDataPoint.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+    public long getDuration() {
+        return mDuration;
+    }
+
+    public long getStartMs() {
+        return mStartMs;
+    }
+
+    public long getmIOBytes(){
```
Comment (jiacheliu3) [inline dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchDataPoint.java]: can you change this to `getIOBytes()`?

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java [hunk 13]
```diff
@@ -12,12 +12,12 @@
...
+import alluxio.stress.StressConstants;
 import alluxio.stress.TaskResult;
+import alluxio.util.FormatUtils;
+import org.HdrHistogram.Histogram;
 
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (jiacheliu3) [inline dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java]: don't use `import *`, use explicit import instead
```
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
```
https://blog.marcnuri.com/intellij-idea-how-to-disable-wildcard-imports

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java (hunk 3 above)
Comment (jiacheliu3) [reply dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java]: let's remove this line for now

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java [hunk 14]
```diff
@@ -155,11 +161,42 @@ public void setErrors(List<String> errors) {
...
 
+  public List<Long> getDurationPercentiles() {
+    return mDurationPercentiles;
+  }
+
+  public void setDurationPercentiles(List<Long> percentiles) {
+    mDurationPercentiles = percentiles;
+  }
+
+  public void generatePercentiles() {
```
Comment (jiacheliu3) [inline dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java]:
```suggestion
  public void calculatePercentiles() {
```

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java [hunk 15]
```diff
@@ -155,11 +161,42 @@ public void setErrors(List<String> errors) {
...
+    Histogram durationHistogram = new Histogram(
+            FormatUtils.parseTimeSize(mParameters.mDuration)
+                    + FormatUtils.parseTimeSize(mParameters.mWarmup),
```
Comment (jiacheliu3) [inline dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java]: why do you add warm up here?

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java (hunk 3 above)
Comment (twalluxio) [reply dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java]: removed for now

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java (hunk 13 above)
Comment (twalluxio) [reply dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java]: caused by addition of java.util.Collection; explicit imports declared

File: dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java (hunk 15 above)
Comment (twalluxio) [reply dora/stress/common/src/main/java/alluxio/stress/worker/WorkerBenchTaskResult.java]: This parameter defines the biggest possible value of the histogram. Since we haven't removed the warmup step, it is still possible for one file-read process to be longer than the entire duration (when duration is very small). But it won't be a problem to set the value as the duration, I will fix it.

**Review by jiacheliu3 — APPROVED** [review]
LGTM, thanks for the work!
